test/modules/test_dangling_dns_detector.py

```diff
@@ -0,0 +1,263 @@
+from test.base import ArtemisModuleTestCase
+from typing import Any
+from unittest.mock import MagicMock, patch
+
+import dns.name
+from dns import rdatatype
+from karton.core import Task
+
+from artemis.binds import TaskStatus, TaskType
+from artemis.modules.dangling_dns_detector import DanglingDnsDetector
+
+
+class TestDanglingDnsDetector(ArtemisModuleTestCase):
+    karton_class = DanglingDnsDetector  # type: ignore
+
+    @patch("artemis.modules.dangling_dns_detector.direct_dns_query")
+    @patch("artemis.modules.dangling_dns_detector.edns_query")
+    @patch("dns.resolver.resolve")
+    def test_check_ns_dangling(self, mock_resolve, mock_edns_query, mock_direct_dns_query) -> None:  # type: ignore
+        # given
+        ns_record = MagicMock()
+        ns_record.rdtype = rdatatype.NS
+        ns_record.target.to_text.return_value = "ns1.unregistered.example.com."
+        mock_ns_answer = MagicMock()
+        mock_ns_answer.rrset = [ns_record]
+        mock_ns_answer.qname = dns.name.from_text("dangling.example.com.")
+        mock_ns_answer.__iter__.return_value = iter([ns_record])
+        mock_resolve.side_effect = [mock_ns_answer]
+        # edns_query returns None for both A and AAAA
+        mock_edns_query.side_effect = [None, None]
+
+        # when
+        result: list[dict[str, Any]] = []
+        self.karton.check_ns("dangling.example.com", result)
+
+        # then
+        self.assertTrue(len(result) == 1)
+        self.assertTrue(result[0]["record"] == rdatatype.NS)
+        self.assertTrue("dangling NS" in result[0]["message"])
+
+    @patch("artemis.modules.dangling_dns_detector.direct_dns_query")
+    @patch("artemis.modules.dangling_dns_detector.edns_query")
+    @patch("dns.resolver.resolve")
+    def test_check_ns_valid(self, mock_resolve, mock_edns_query, mock_direct_dns_query) -> None:  # type: ignore
+        # given
+        ns_record = MagicMock()
+        ns_record.rdtype = rdatatype.NS
+        ns_record.target.to_text.return_value = "ns1.valid.example.com."
+        mock_ns_answer = MagicMock()
+        mock_ns_answer.rrset = [ns_record]
+        mock_ns_answer.qname = dns.name.from_text("valid.example.com.")
+        mock_ns_answer.__iter__.return_value = iter([ns_record])
+        mock_resolve.side_effect = [mock_ns_answer]
+
+        a_record = MagicMock()
+        a_record.rdtype = rdatatype.A
+        a_record.address = "203.0.113.10"
```

Comment:
is this a specific IP? what about something like 8.8.8.8?

Reply:
It's mocked at the end of the day, but I've changed test to 8.8.8.8 so it can be better understood.